### ISSUE TITLE
Generalized integrator class for use with FMUs

### ISSUE THREAD

**Issue by jaeandersson** [casadi/casadi#2823]
The binary FMU interface (#2779) allows FMUs to be evaluated pointwise. This is would in principle allow the FMU to be integrated using the existing Integrator class, but this would potentially be very limited and inefficient, due to:
 * Excessive initializations of the FMU, which is expensive
 * No event handling
 * No easy way to incorporate piecewise constant control inputs
 * Derivative calculation might be very expensive
 * Need to keep track of which states are differential and which are algebraic

To get around this, it makes sense to add a dedicated Simulator plugin class which can be constructed from an FMU function instance:
```python
dae = DaeBuilder(...)
...
f = dae.fmu_fun(...)
time_grid = ...
simulator = casadi.simulator('simulator', f, time_grid)
```
To start with, the simulator class would only be defined for fmu_functions, but it could easily be extended to symbolic representations of the dynamic system. It would also not be differentiable, at least not at first. It might be possible and desirable to combine this class with the integrator class, but given the disruption it would result in, it's better to keep the classes separate at first, but with maximum amount of communality. The dynamic system argument should correspond to a state-space representation with quadratures: `(t, x, z, p, u) -> (xdot, alg, y, qdot)`. The Simulator class will have the same signature as Integrator but add piecewise controls ("u") and outputs ("y").

To be able to simulate an FMU, a CVodes plugin makes sense to begin with.

*Update:*
This issue has taken some unexpected turns. The latest conclusion is that maintaining both `Integrator` and `Simulator` is not realistic. Instead, `Integrator` has now been extended with elements from the `Simulator` class, in particular control inputs. The output support has been left out since they can be calculated after the integration with few drawbacks. What remains of this issue is to port forward sensitivity approach from `Simulator` to `Integrator`, i.e. add "nfwd" as an option, instead of deducing it from the number of columns in the oracle_ inputs/outputs.

**Comment by jaeandersson** [comment]
Closing this issue. Additional features handled in dedicated issues.

**Comment by jaeandersson** [comment]
Reopened to better align syntax with integrator, and to add quadrature support

**Comment by jaeandersson** [comment]
Work completed. This issue is a major change internally, but with limited changes for the end-user. Relevant changes for the end-user include:
 * Deprecation of the `t0`, `tf`, `output_t0` and `grid` options. They are now replaced by additional, optional, constructor arguments.
 * An additional `nfwd` argument to obtain forward sensitivity equations as additional right-hand-sides in the outputs.
 * Breaking changes to the serialization API for the integrator class specifically.

**Comment by Rebew111** [comment]
Hei All,

Is it possible to upload a simple optimal control example from fmu model? That would definitely be very helpful for my research. 

Kind regards Simon Weber

**Comment by jaeandersson** [comment]
> Hei All,
> 
> Is it possible to upload a simple optimal control example from fmu model? That would definitely be very helpful for my research.
> 
> Kind regards Simon Weber

Sorry, things got a bit hectic before the release and the FMU demo I had never got uploaded. You can find the latest version here:
https://github.com/casadi/casadi/blob/7e20560edaa22aedffa83320428d1d4615562998/docs/notebooks/fmu_demo.ipynb

The demo is pretty rough, and I just discovered that one of part of it got accidentally broken before the 3.6.0 release. I'll try to include a fixed version in 3.6.1
